### ISSUE TITLE
Form: RepeatedType Field ignores validation if single value is passed

### ISSUE THREAD

**Issue by HKandulla** [symfony/symfony#21242]
| Q                | A
| ---------------- | -----
| Bug report?      | yes
| Feature request? | no
| BC Break report? | no
| RFC?             | no
| Symfony version  | 3.2.1

I came across a strange behaviour that looks like a bug to me. We use Symfony for a JSON API and fill the forms with arrays. When I submit an array where the password consists of an array with two keys, the validation workd fine. But when pass an array where password has a single value or even an empty string, the validation is ignored and the form validates.

```
[
	"password": [
		"first": "string",
		"second": "string"
	]
]
```
validates according to rules

```
[
	"password": ""
] 
```
ignores validation rules and validates to true

FormType
```
class UserPasswordType extends AbstractType
{
    /**
     * @param FormBuilderInterface $builder
     * @param array                $options
     */
    public function buildForm(FormBuilderInterface $builder, array $options)
    {
        $builder
            ->add('password', RepeatedType::class, array(
                'required' => true,
                'type' => PasswordType::class,
                'invalid_message' => 'The password fields must match.',
                ))
        ;
    }

    /**
     * @param OptionsResolver $resolver
     */
    public function configureOptions(OptionsResolver $resolver)
    {
        $resolver->setDefaults(array(
            'data_class' => User::class,
            'csrf_protection' => false,
        ));
    }

    /**
     * @return string
     */
    public function getName()
    {
        return 'app_user_password';
    }
}
```

Action
```
    public function updatePasswordAction(User $user, Request $request)
    {
        $data = $this->getDataArrayFromRequest($request);
        $form = $this->createForm(UserPasswordType::class, $user);
        $form->submit($data);
        if ($form->isValid()) {
            $this->get('app.entity.user_manager')->updatePassword($user);

            return $this->userJsonResponse($user, 201);
        }

        return $this->formErrorsJsonResponse($form);
    }
```

**Comment by jakzal** [comment]
You didn't really make your password field required. Setting the `required` option to `true` [only enables client-side html5 validation](http://symfony.com/doc/current/reference/forms/types/form.html#required). You'll need to define [a list of constraints](http://symfony.com/doc/current/reference/forms/types/form.html#constraints) to validate it on the server side. 

Closing as this is not a bug.

**Comment by HeahDude** [comment]
Hey @HKandulla thanks for opening this issue.

I think this is expected here, since the default validation of the repeated type only check that both values are the same.

To support your case you should use a `NotBlank` constraint, since the submitted empty string is handled as `null` converted in this case to an empty array because the type is compound.

**Comment by HKandulla** [comment]
Hi @jakzal, @HeahDude thanks for your quick responses and help!

I added the constraints to the entity properties and thought the form would validate against those (which it does if the field is present/submitted, I guess).

So if I understand you/forms right, the form does validate against constraints defined on the entity properties, but only if the field (key) is present in the submitted data. So, even if I have complex constraints defined on the entity properties, it is enough to have a NotBlank constraint defined on form field, because once the field is present, the constraints defined on the properties would be validated against. Is that correct?

If yes, wouldn't it make sense to trigger this behaviour with "required" => true?

Again, many thanks for your help! Hannes

**Comment by HKandulla** [comment]
Hi @jakzal, @HeahDude,

looks like my theory is wrong. Even with NotBlank and Length(array('min' => 3)) constraints added to the form field. I can still pass and single string as password and the form validates fine (and ignores the value). Looks like I have to make sure that password contains an array with two keys before passing it to the form..

´´´
        $builder
            ->add('password', RepeatedType::class, array(
                'type' => PasswordType::class,
                'invalid_message' => 'The password fields must match.',
                'constraints' => array(
                    new NotBlank(),
                    new Length(array('min' => 3)),
                )
            ))
        ;
´´´
the form doesn't throw an validation error if I pass and single empty string

**Comment by jakzal** [comment]
Have you tried with [first_options & second_options](https://symfony.com/doc/current/reference/forms/types/repeated.html#first-options)?

**Comment by HKandulla** [comment]
unfortunately, first_options and second_options doesn't help. If password is of type array it gets validated if it is of type string it gets ignored and validation passes.

**Comment by HeahDude** [comment]
Could you please try to add `'empty_data' => array()` to the password field?

**Comment by HKandulla** [comment]
`'empty_data' => array()` doesn't help either.

**Comment by HKandulla** [comment]
I added all suggestions together:

```
    public function buildForm(FormBuilderInterface $builder, array $options)
    {
        $builder
            ->add('password', RepeatedType::class, array(
                'type' => PasswordType::class,
                'invalid_message' => 'The password fields must match.',
                'constraints' => array(
                    new NotBlank(),
                    new Length(array('min' => 3)),
                ),
                'first_options' => array(
                    'constraints' => array(
                        new NotBlank(),
                        new Length(array('min' => 3)),
                    ),

                ),
                'second_options' => array(
                    'constraints' => array(
                        new NotBlank(),
                        new Length(array('min' => 3)),
                    ),
                ),
                'empty_data' => array(),
            ))
        ;
    }
```

But 
```
[
	"password": ""
]
```
still validates to true

**Comment by xabbuh** [comment]
Would you be able to provide a failing test case that illustrates your issue? Maybe fork the Symfony Standard Edition and add a functional test.

**Comment by ogizanagi** [comment]
I was able to reproduce the issue when submitting `['password': ""]`.
It seems a `TransformationFailedException` is thrown with `#message: "Compound forms expect an array or NULL on submission."`. But the fact is that for some reasons, no error is added to the "password" compound form...

I think #20966 would solve it, at least by not considering the form valid. But still not having the error mapped is weird. :/

Status: Reviewed

**Comment by ogizanagi** [comment]
To complete: The error mapping does not happen, because the `error_mapping` option of the form is set to `['.' => 'first]` by the [RepeatedTypeValidatorExtension](https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Form/Extension/Validator/Type/RepeatedTypeValidatorExtension.php#L28-L34).
But as the `form[password][first]` is not submitted (`form[password]`is instead, with a string value), the [`ViolationMapper::acceptsErrors()`](https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Form/Extension/Validator/ViolationMapper/ViolationMapper.php#L126) method will return `false` and won't map anything. The `TransformationFailedException` is vanished. 🍃

**Comment by HeahDude** [comment]
Thanks @ogizanagi, you're right, I'm working on it ;)

**Comment by HKandulla** [comment]
Thanks for working on this. If I can help out, please let me know.

**Comment by pedrocasado** [comment]
Any updates here? Workaround? I'm on the same situation..

**Comment by dohpaz42** [comment]
@HeahDude ,

If it helps, I found a work-around for this issue by adding an EventSubscriber to the RepeatedType field:

    <?php
    
    namespace AppBundle\Form\Extension\EventListener;
    
    use Symfony\Component\EventDispatcher\EventSubscriberInterface;
    use Symfony\Component\Form\FormEvent;
    use Symfony\Component\Form\FormEvents;
    
    class FixRepeatedTypeListener implements EventSubscriberInterface
    {
        /**
         * {@inheritdoc}
         */
        public static function getSubscribedEvents()
        {
            return [FormEvents::PRE_SUBMIT => 'preSubmit'];
        }
    
        public function preSubmit(FormEvent $event)
        {
            if ($event->getForm()->getConfig()->getType()->getName() !== 'repeated') {
                throw new \UnexpectedValueException(sprintf(
                    'Expected FormType of type "repeated", "%s" given',
                    $event->getForm()->getConfig()->getType()->getName()
                ));
            }
    
            if (!is_array($event->getData())) {
                $event->setData(['first' => $event->getData(), 'second' => null]);
            }
        }
    }


In short, we coerce the RepeatedType field's submitted value into the expected array format so that validation can be triggered.

HTH! And thank you!

**Comment by geoff-maddock** [comment]
Having a similar issue to this - when I use RepeatedType in a form with a data class equal to an entity, and submit empty values, it ignores all asserts in the entity definition.   If i enter anything in either field, the asserts are checked.   

For the time being, I'm able to add a constraint into the first_options config that ends up being checked even when nothing is submitted in either field.  But I don't believe I should have to do that.

Ex:
```
'first_options' => [
                'label' => 'Password',
                'attr' => ['autofocus' => 'autofocus'],
                'required' => true,
                'constraints' => array(
                    new NotBlank(array('groups' => 'password')),
                )
```

**Comment by xabbuh** [comment]
Can you confirm that this will be fixed by #28731?

**Comment by evgenyvas** [comment]
Version 4.2.1 - bug is still exists.
Constraint in data class is ignored. Workaround with manually setting constraints option for 'repeated' form widget is working ok.

**Comment by xabbuh** [comment]
@evgenyvas Can you open a new issue and best provide a small example application that allows to reproduce your issue?

**Comment by evgenyvas** [comment]
It was my mistake, wrong mapping form to data object
```
    public function mapFormsToData($forms, &$data) {
        $forms = iterator_to_array($forms);
        $data = new \App\DataValues\UserData(
            $forms['accountNumber']->getData(),
            $forms['city']->getData(),
            $forms['street']->getData(),
            $forms['house']->getData(),
            $forms['housing']->getData(),
            $forms['room']->getData(),
            $forms['contractNumber']->getData(),
            $forms['name']->getData(),
            $forms['email']->getData(),
            $forms['phone']->getData(),
            null,//$forms['roles']->getData(),
            $forms['password']->getData()
        );
    }
```
